src/pybind/mgr/rgw/module.py

```diff
@@ -4,28 +4,55 @@
 import errno
 import base64
 import functools
+import sys
 
-from mgr_module import MgrModule, CLICommand, HandleCommandResult
+from mgr_module import MgrModule, CLICommand, HandleCommandResult, Option
 import orchestrator
 
 from ceph.deployment.service_spec import RGWSpec, PlacementSpec, SpecValidationError
-from typing import Any, Optional, Sequence, Iterator, List
+from typing import Any, Optional, Sequence, Iterator, List, Callable, TypeVar, cast, Dict, Tuple, Union
 
 from ceph.rgw.types import RGWAMException, RGWAMEnvMgr, RealmToken
 from ceph.rgw.rgwam_core import EnvArgs, RGWAM
-from orchestrator import OrchestratorClientMixin, OrchestratorError
+from orchestrator import OrchestratorClientMixin, OrchestratorError, DaemonDescription, OrchResult
+
+
+FuncT = TypeVar('FuncT', bound=Callable[..., Any])
+
+# this uses a version check as opposed to a try/except because this
+# form makes mypy happy and try/except doesn't.
+if sys.version_info >= (3, 8):
+    from typing import Protocol
+else:
+    # typing_extensions will not be available for the real mgr server
+    from typing_extensions import Protocol
+
+
+class MgrModuleProtocol(Protocol):
```

Comment:
For my info, what class is this protocol class meant to be modeling? I see the  `RGWAMOrchMgr` class has all three of the same functions, but all the return types are different. It seems like the type is being applied to the `mgr` var passed to the `RGWAMOrchMgr` `__init__` but that appears to be an instance of the `Module` class which doesn't have any of these functions defined.

Reply:
The thing is `RGWAMOrchMgr` instance is initialized by `RGWAMOrchMgr(self)`. self in this case refers to the `class Module(orchestrator.OrchestratorClientMixin, MgrModule)` which inherits both from `OrchestratorClientMixin` (the orchestrator) and `MgrModule`. As of this multiple inheritance it gets access to the methods `apply_rgw` and `list_daemons` from the orchestrator base class and `tool_exec` from the `MgrModule` base class.  In order to indicate to mypy that this class has access to all these methods (otherwise mypy checks fail) we create the protocol class `MgrModuleProtocol`.